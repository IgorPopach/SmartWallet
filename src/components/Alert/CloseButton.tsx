import React from 'react';

interface Props {
    onClose: () => void;
    classes: string;
}

export const CloseButton = ({ onClose, classes }: Props) => {
    return (
        <button type="button" className={classes} onClick={onClose}>
            <span aria-hidden="true">&times;</span>
        </button>
    );
};
