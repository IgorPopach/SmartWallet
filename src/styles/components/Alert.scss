.alert {
    margin-bottom: .5rem;
    padding: 5px 15px;
    position: relative;

    button {
        position: absolute;
        top: 0;
        right: 4px;
        outline: none;

        &.close {
            font-weight: 400;
        }
    }

    &-show {
        animation: show-message 1s;
        
    }
    &-hide {
        animation: hide-message 1s;
        transition-property: border, height, overflow, padding, margin; 
        transition-duration: 1.4s; 
        transition-timing-function: ease;
        border: none;
        height: 0;
        overflow: hidden;
        padding: 0;
        margin: 0
    }
}

@keyframes show-message {
    0% {top: -300px}
    80% {top: 10px}
    100% {top: 0}
}

@keyframes hide-message {
    0% {top: 0}
    15% {top: 15px}
    100% {top: -500px}
}
