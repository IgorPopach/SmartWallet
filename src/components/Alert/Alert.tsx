import React from 'react';

import { Message } from '../../store/snackbar/types';
import { CloseButton } from './CloseButton';

import './../../styles/components/Alert.scss';

interface Props {
    message: Message;
    onClose?: () => void;
    delay?: number;
}

function selfClose(closeAction: Props['onClose'], delay: number) {
    if (closeAction && delay) {
        return window.setTimeout(closeAction, delay + 1500);
    }
    return 0;
}

function playAnimation(setAnimationStyle: (style: string) => void, delay: number) {
    if (delay) {
        return window.setTimeout(() => {
            setAnimationStyle('hide');
        }, delay);
    }
    return 0;
}

const Alert = ({ message, onClose, delay }: Props) => {
    const [style, setStyle] = React.useState('show');

    const handleClick = () => {
        setStyle('hide');
        setTimeout(() => onClose(), 2000);
    };

    React.useEffect(() => {
        const timerAnimation = playAnimation(setStyle, delay);
        const timerSelfClose = selfClose(onClose, delay);
        return () => {
            clearTimeout(timerSelfClose);
            clearTimeout(timerAnimation);
        };
    }, []);

    return (
        <div className={`alert alert-${message.type} alert-${style}`}>
            <h2>{message.title}</h2>
            <p>{message.text}</p>
            {onClose && <CloseButton onClose={handleClick} />}
        </div>
    );
};

export default Alert;
