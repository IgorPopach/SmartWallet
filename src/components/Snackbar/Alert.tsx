import React, { useCallback } from 'react';

import { Message } from './../../store/snackbar/types';

import './../../styles/components/Alert';

interface Props {
    message: Message,
    onSubmit: (id: string) => void
}

const Alert: React.FC<Props> = ({ message, onSubmit }) => {

    const submitHandler = () => onSubmit(message.id)

    return (
        <div className={`alert alert-${message.type}`}>
            <p>
                <strong>{message.title}</strong>
            </p>
            <p>
                {message.text}
            </p>
            <button
                type="button"
                className="close"
                onClick={submitHandler}
            >
                <span aria-hidden="true">&times;</span>
            </button>
        </div>
    )
}

export default Alert;