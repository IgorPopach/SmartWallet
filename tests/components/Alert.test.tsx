import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import Alert from './../../src/components/Alert/Alert';

type Props = React.ComponentProps<typeof Alert>;

describe('<Alert />', () => {
    const props: Props = {
        message: {
            id: '1',
            title: 'Test alert 1',
            messageText: 'this is test text for Alert',
            type: 'info',
        },
        delay: 3000,
    };

    let wrapper: ShallowWrapper<Props>;

    beforeAll(() => {
        wrapper = shallow(<Alert {...props} />);
    });

    it('should have correct layout', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
