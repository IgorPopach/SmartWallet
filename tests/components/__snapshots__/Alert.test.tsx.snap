// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should have correct layout 1`] = `
<div
  className="alert alert-info alert-show"
>
  <h2>
    Test alert 1
  </h2>
  <p>
    this is test text for Alert
  </p>
</div>
`;
